crates/turborepo-lib/src/process/child.rs

```diff
@@ -399,60 +399,44 @@ impl Child {
     /// single writers.
     pub async fn wait_with_single_piped_output<W: Write>(
         &mut self,
-        pipe: W,
+        mut pipe: W,
     ) -> Result<Option<ChildExit>, std::io::Error> {
-        // Note this is similar to tokio::process::Command::wait_with_outputs
-        // but allows us to provide our own sinks instead of just writing to a buffers.
-        async fn pipe_lines<R: AsyncRead + Unpin>(
-            stream: Option<R>,
-            tx: mpsc::Sender<Vec<u8>>,
-        ) -> std::io::Result<()> {
-            let Some(stream) = stream else { return Ok(()) };
-            let stream = BufReader::new(stream);
-            let mut lines = AsyncByteLines::new(stream);
-            while let Some(line) = lines.next().await? {
-                let line = {
-                    // Allocate vector with enough capacity for trailing newline
-                    let mut buffer = Vec::with_capacity(line.len() + 1);
-                    buffer.extend_from_slice(line);
-                    // Line iterator doesn't return newline delimiter so we must add it here
-                    buffer.push(b'\n');
-                    buffer
-                };
-
-                if tx.send(line).await.is_err() {
-                    // If the receiver is dropped then we have nothing to do with these bytes
-                    break;
-                }
+        async fn next_line<R: AsyncBufRead + Unpin>(
+            stream: &mut Option<R>,
+            buffer: &mut Vec<u8>,
+        ) -> Option<Result<(), io::Error>> {
+            match stream {
+                Some(stream) => match stream.read_until(b'\n', buffer).await {
+                    Ok(0) => None,
+                    Ok(_) => Some(Ok(())),
+                    Err(e) => Some(Err(e)),
+                },
+                None => None,
             }
-            Ok(())
         }
 
-        async fn write_lines<W: Write>(
-            mut rx: mpsc::Receiver<Vec<u8>>,
-            mut writer: W,
-        ) -> std::io::Result<()> {
-            while let Some(buffer) = rx.recv().await {
-                writer.write_all(&buffer)?;
+        let mut stdout_lines = self.stdout().map(BufReader::new);
+        let mut stderr_lines = self.stderr().map(BufReader::new);
+
+        let mut stdout_buffer = Vec::new();
+        let mut stderr_buffer = Vec::new();
+        loop {
+            tokio::select! {
+                Some(result) = next_line(&mut stdout_lines, &mut stdout_buffer) => {
```

Comment:
If I'm understanding the [`AsyncBufReadExt`](https://docs.rs/tokio/latest/tokio/io/trait.AsyncBufReadExt.html) docs correctly, there is a *slight* race condition here. On cancelation, the partially read bytes are written to the buffer and the bytes written counter is reset. So it's possible that both stdout and stderr are completely written to the buffer, one of them wins, which leaves the other with a full buffer, but its counter is reset to 0, so the next time around it gets polled, returns a `None` and gets disabled. 

In which case, we should write to pipe one last time in the else branch.

Reply:
Good catch!